deso/index.go

```diff
@@ -24,171 +24,238 @@ const (
 	if err != nil {
 		return nil, err
 	}
-
-	opBundle := &lib.UtxoOperationBundle{}
-
-	err = index.db.View(func(txn *badger.Txn) error {
-		utxoOpsItem, err := txn.Get(index.utxoOpsKey(blockHash))
-		if err != nil {
-			return err
-		}
-
-		return utxoOpsItem.Value(func(valBytes []byte) error {
-			rr := bytes.NewReader(valBytes)
-			if exist, err := lib.DecodeFromBytes(opBundle, rr); !exist || err != nil {
-				return errors.Wrapf(err, "Problem decoding utxoops, exist: (%v)", exist)
-			}
-			return nil
-		})
-	})
-	if err != nil {
-		return nil, err
-	}
-
-	return opBundle.UtxoOpBundle, nil
+	return lib.GetUtxoOperationsForBlock(index.chainDB, index.snapshot, blockHash)
 }
 
 //
 // Balance Snapshots
 //
 
-func balanceSnapshotKey(isLockedBalance bool, publicKey *lib.PublicKey, blockHeight uint64, balance uint64) []byte {
-	startPrefix := PrefixBalanceSnapshots
-	if isLockedBalance {
-		startPrefix = PrefixLockedBalanceSnapshots
+type BalanceType uint8
+
+const (
+	DESOBalance                BalanceType = 0
+	CreatorCoinLockedBalance   BalanceType = 1
+	ValidatorStakedDESOBalance BalanceType = 2
+	LockedStakeDESOBalance     BalanceType = 3
+)
+
+func (balanceType BalanceType) BalanceTypePrefix() byte {
+	switch balanceType {
+	case DESOBalance:
+		return PrefixBalanceSnapshots
+	case CreatorCoinLockedBalance:
+		return PrefixLockedBalanceSnapshots
+	case ValidatorStakedDESOBalance:
+		return PrefixValidatorStakedBalanceSnapshots
+	case LockedStakeDESOBalance:
+		return PrefixLockedStakeBalanceSnapshots
+	default:
+		panic("unknown balance type")
 	}
+}
 
-	prefix := append([]byte{}, startPrefix)
+func balanceSnapshotKey(balanceType BalanceType, publicKey *lib.PublicKey, blockHeight uint64, balance uint64) []byte {
+	prefix := append([]byte{}, balanceType.BalanceTypePrefix())
 	prefix = append(prefix, publicKey[:]...)
 	prefix = append(prefix, lib.EncodeUint64(blockHeight)...)
 	prefix = append(prefix, lib.EncodeUint64(balance)...)
 	return prefix
 }
 
-func hypersyncHeightToBlockKey(blockHeight uint64, isLocked bool) []byte {
+func lockedStakeBalanceSnapshotKey(
+	balanceType BalanceType, stakerPKID *lib.PKID, validatorPKID *lib.PKID, lockedAtEpochNumber uint64,
+	blockHeight uint64, balance uint64,
+) []byte {
+	prefix := append([]byte{}, balanceType.BalanceTypePrefix())
+	prefix = append(prefix, stakerPKID[:]...)
+	prefix = append(prefix, validatorPKID[:]...)
+	prefix = append(prefix, lib.EncodeUint64(lockedAtEpochNumber)...)
+	prefix = append(prefix, lib.EncodeUint64(blockHeight)...)
+	prefix = append(prefix, lib.EncodeUint64(balance)...)
+	return prefix
+}
 
-	lockedByte := byte(0)
-	if isLocked {
-		lockedByte = byte(1)
+func (balanceType BalanceType) HypersyncBalanceTypePrefix() byte {
+	switch balanceType {
+	case DESOBalance:
+		return 0
+	case CreatorCoinLockedBalance:
+		return 1
+	case ValidatorStakedDESOBalance:
+		return 2
+	case LockedStakeDESOBalance:
+		return 3
+	default:
+		panic("unknown balance type")
 	}
+}
```

Comment:
Nit: I think this would work just fine and be less "magic number"-ey if we replaced it with:
`return byte(balanceType.BalanceTypePrefix())`. So the Hypersync prefixes are actually just: `{PrefixHypersyncBlockHeightToBalances, whatever the balance prefix is}`.

Whole function would look like this:
```
func (balanceType BalanceType) HypersyncBalanceTypePrefix() byte {
	return byte(balanceType.BalanceTypePrefix())
}
```

Reply:
Actually you could just delete this function and use `byte(balanceType.BalanctTypePrefix())` in its place.

Reply:
No we can't use `byte(balanceType.BalanceTypePrefix())` as this breaks backwards compatibility. We can use `byte(balanceType)` if you prefer.